improved parse of stream data

yusuga opened the pull request:
The stream data might be acquired intermittently.

nst commented:
Thank you for these improvements!

But shouldn't we handle all possible message types?

https://dev.twitter.com/streaming/overview/messages-types

yusuga commented:
Thank you for reply.

> But shouldn't we handle all possible message types?
> https://dev.twitter.com/streaming/overview/messages-types

I added some convenience stream methods.

---

However, this new parsing process will need to be added.
In the current parsing process, there is Tweet JSON that missed.

Twitter streaming does not necessarily able to get the full JSON string. Thus, it is necessary to perform SAX-style parsing.(such as [YAJL](https://github.com/gabriel/yajl-objc))
The new persing process is performed a SAX-style parsing using 'delimited = @"length"'.

In the following way, you can confirm that the json string data has been divided.

## In Current version
### POST

```objectivec
[_twitter postStatusesFilterKeyword:@"is"
                          progressBlock:^(NSDictionary *tweet) {                              
                          } errorBlock:^(NSError *error) {                              
                          }];
```

### Parse process (Added parse error log)

```objectivec
- (void)handleStreamingResponse:(NSHTTPURLResponse *)urlResponse request:(id)request data:(NSData *)responseData {
    
    if(responseData == nil) {
        self.errorBlock(request, [self requestHeadersForRequest:request], [urlResponse allHeaderFields], nil);
        return;
    }
    
    NSError *jsonError = nil;
    NSJSONSerialization *json = [NSJSONSerialization JSONObjectWithData:responseData options:NSJSONReadingAllowFragments error:&jsonError];
    
    if([json valueForKey:@"error"]) {
        
        NSString *message = [json valueForKey:@"error"];
        NSDictionary *userInfo = [NSDictionary dictionaryWithObject:message forKey:NSLocalizedDescriptionKey];
        NSError *jsonErrorFromResponse = [NSError errorWithDomain:NSStringFromClass([self class]) code:0 userInfo:userInfo];
        
        self.errorBlock(request, [self requestHeadersForRequest:request], [urlResponse allHeaderFields], jsonErrorFromResponse);
        
        return;
    }
    
    // we can receive several dictionaries in the same data chunk
    // such as '{..}\r\n{..}\r\n{..}' which is not valid JSON
    // so we split them up into a 'jsonChunks' array such as [{..},{..},{..}]
    
    NSString *jsonString = [[NSString alloc] initWithData:responseData encoding:NSUTF8StringEncoding];
    NSArray *jsonChunks = [jsonString componentsSeparatedByString:@"\r\n"];
    
    for(NSString *jsonChunk in jsonChunks) {
        if([jsonChunk length] == 0) continue;
        NSData *data = [jsonChunk dataUsingEncoding:NSUTF8StringEncoding];
        NSError *jsonError = nil;
        id json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&jsonError];
        if(json) {
            self.completionBlock(request, [self requestHeadersForRequest:request], [urlResponse allHeaderFields], json);
        } else {
            static NSUInteger count;
            printf("----------------------------------------------------------------- (%zd)\n", count++);
            printf("@%s\n\n", [jsonChunk cStringUsingEncoding:NSUTF8StringEncoding]);
        }
    }
}
```

### Log (Data has been acquired by dividing)

```objectivec
----------------------------------------------------------------- (0)
@{"created_at":"Tue Mar 24 05:34:15 +0000 2015","id":580241191553216512,"id_str":"580241191553216512","text":"Really craving superman ice-cream. The struggle is real.","source":"\u003ca href=\"http:\/\/twitter.com\/download\/iphone\" rel=\"nofollow\"\u003eTwitter for iPhone\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"user":{"id":2752250897,"id_str":"2752250897","name":"Mackenzie Schumann","screen_name":"MKSchumann","location":"","url":null,"description":"Air Force Wife. 18. Twin. Lover. Fighter. Survivor.","protected":false,"verified":false,"followers_count":74,"friends_count":207,"listed_count":0,"favourites_count":142,"statuses_count":245,"created_at":"Wed Aug 27 03:23:47 +0000 2014","utc_offset":-14400,"time_zone":"Eastern Time (US & Canada)","geo_enabled":false,"lang":"en","contributors_enabled":false,"is_translator":false,"profile_background_color":"C0DEED","profile_background_image_url":"http:\/\/abs.twimg.com\/images\/themes\/theme1\/bg.png","profile_background_image_url_https":"https:\/\/abs.twimg.com\/images\/themes\/theme1\/bg.pn

----------------------------------------------------------------- (1)
@g","profile_background_tile":false,"profile_link_color":"0084B4","profile_sidebar_border_color":"C0DEED","profile_sidebar_fill_color":"DDEEF6","profile_text_color":"333333","profile_use_background_image":true,"profile_image_url":"http:\/\/pbs.twimg.com\/profile_images\/532393303648395264\/3dyErNcm_normal.jpeg","profile_image_url_https":"https:\/\/pbs.twimg.com\/profile_images\/532393303648395264\/3dyErNcm_normal.jpeg","profile_banner_url":"https:\/\/pbs.twimg.com\/profile_banners\/2752250897\/1414827512","default_profile":true,"default_profile_image":false,"following":null,"follow_request_sent":null,"notifications":null},"geo":null,"coordinates":null,"place":null,"contributors":null,"retweet_count":0,"favorite_count":0,"entities":{"hashtags":[],"trends":[],"urls":[],"user_mentions":[],"symbols":[]},"favorited":false,"retweeted":false,"possibly_sensitive":false,"filter_level":"low","lang":"en","timestamp_ms":"1427175255083"}

----------------------------------------------------------------- (2)
@{"created_at":"Tue Mar 24 05:34:15 +0000 2015","id":580241194258599936,"id_str":"580241194258599936","text":"Kate Mulgrew is on this American Dad. Hey Kate! \noh and Stan Lee! Hi Stan!","source":"\u003ca href=\"http:\/\/twitter.com\" rel=\"nofollow\"\u003eTwitter Web Client\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"user":{"id":80948977,"id_str":"80948977","name":"Joe","screen_name":"thisjoeiknow","location":"i don't go outside","url":null,"description":"Kiss Me, I'm Garbage.","protected":false,"verified":false,"followers_count":244,"friends_count":454,"listed_count":5,"favourites_count":12202,"statuses_count":14031,"created_at":"Thu Oct 08 21:31:53 +0000 2009","utc_offset":-18000,"time_zone":"Central Time (US & Canada)","geo_enabled":true,"lang":"en","contributors_enabled":false,"is_translator":false,"profile_background_color":"7A5876","profile_background_image_url":"http:\/\/pbs.twimg.com\/profile_background_images\/321844021\/lumpy_space_princess_by_ericlovesdonna-d38netr.png","profile_background_image_url_https":"https:\/\/pbs.twimg.com\/profile_background_images\/321844021\/lumpy_space_princess_by_ericlovesdonna-d38netr.png","profile_background_tile":false,"profile_link_color":"3737B8","profile_sidebar_border_color":"000000","profile_sidebar_fill_color":"9CE0E6","profile_text_color":"333333","profile_use_background_image":false,"profile_image_url":"http:\/\/pbs.twimg.com\/profile_images\/578067699978272768\/QzJwJ1Ba_normal.jpeg","profile_image_url_https":"https:\/\/pbs.twimg.com\/profile_images\/578067699978272768\/QzJwJ1Ba_normal.jpeg","profile_banner_url":"https:\/\/pbs.twimg.com\/profile_banners\/80948977\/1404260678","default_profile":false,"default_profile_image":false,"following":null,"follow_request_sent":null,"notifications":null},"geo":null,"coordinates":null,"place":null,"contributors":null,"retweet_count"

----------------------------------------------------------------- (3)
@:0,"favorite_count":0,"entities":{"hashtags":[],"trends":[],"urls":[],"user_mentions":[],"symbols":[]},"favorited":false,"retweeted":false,"possibly_sensitive":false,"filter_level":"low","lang":"en","timestamp_ms":"1427175255728"}

```
